server/src/main/java/io/druid/server/coordinator/rules/LoadRule.java

```diff
@@ -50,187 +52,267 @@
-        break;
+    final ReplicationThrottler throttler = params.getReplicationManager();
+    for (int numAssigned = 0; numAssigned < numToAssign; numAssigned++) {
+      if (!throttler.canCreateReplicant(tier)) {
+        return numAssigned;
       }
 
-      final ServerHolder holder = strategy.findNewSegmentHomeReplicator(segment, serverHolderList);
-
+      final ServerHolder holder = params.getBalancerStrategy().findNewSegmentHomeReplicator(segment, holders);
       if (holder == null) {
         log.warn(
             "Not enough [%s] servers or node capacity to assign segment[%s]! Expected Replicants[%d]",
-            tier,
-            segment.getIdentifier(),
-            expectedReplicantsInTier
-        );
-        break;
-      }
-
-      if (replicate) {
-        replicationManager.registerReplicantCreation(
-            tier, segment.getIdentifier(), holder.getServer().getHost()
+            tier, segment.getIdentifier(), targetReplicantsInTier
         );
+        return numAssigned;
       }
+      holders.remove(holder);
 
-      holder.getPeon().loadSegment(
-          segment,
-          new LoadPeonCallback()
-          {
-            @Override
-            public void execute()
-            {
-              replicationManager.unregisterReplicantCreation(
-                  tier,
-                  segment.getIdentifier(),
-                  holder.getServer().getHost()
-              );
-            }
-          }
-      );
-
-      ++assignedCount;
-      ++currReplicantsInTier;
-      ++currTotalReplicantsInCluster;
+      final String segmentId = segment.getIdentifier();
+      final String holderHost = holder.getServer().getHost();
+      throttler.registerReplicantCreation(tier, segmentId, holderHost);
+      holder.getPeon().loadSegment(segment, () -> throttler.unregisterReplicantCreation(tier, segmentId, holderHost));
     }
 
-    return assignedCount;
+    return numToAssign;
   }
 
-  private CoordinatorStats drop(
-      final Map<String, Integer> loadStatus,
+  private static void drop(
+      final Object2IntMap<String> targetReplicants,
+      final Object2IntMap<String> currentReplicants,
+      final DruidCoordinatorRuntimeParams params,
       final DataSegment segment,
-      final DruidCoordinatorRuntimeParams params
+      final CoordinatorStats stats
   )
   {
-    CoordinatorStats stats = new CoordinatorStats();
+    final DruidCluster druidCluster = params.getDruidCluster();
 
     // Make sure we have enough loaded replicants in the correct tiers in the cluster before doing anything
-    for (Integer leftToLoad : loadStatus.values()) {
-      if (leftToLoad > 0) {
-        return stats;
+    for (final Object2IntMap.Entry<String> entry : targetReplicants.object2IntEntrySet()) {
+      final String tier = entry.getKey();
+      // if there are replicants loading in cluster
+      if (druidCluster.hasTier(tier) && entry.getIntValue() > currentReplicants.getOrDefault(tier, 0)) {
```

Comment:
Why if this it true in any tier, we exit the method? maybe still drop segments in other tiers?

Reply:
Yeah, I also do not understand the complete rational behind this decision but the test cases do enforce such a behavior. May be it is to prevent some form of "thrashing" whereby segments get loaded and dropped excessively?

Reply:
@fjy are you able to comment here?

Reply:
@leventov @fjy So should I do any modification for this?